src/perf/multi_reader.cpp

```diff
@@ -19,29 +19,37 @@
  * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#include <lo2s/perf/bio/reader.hpp>
 #include <lo2s/perf/bio/writer.hpp>
+#include <lo2s/perf/io_reader.hpp>
 #include <lo2s/perf/multi_reader.hpp>
 
 namespace lo2s
 {
 namespace perf
 {
-template <class Reader, class Writer>
-MultiReader<Reader, Writer>::MultiReader(trace::Trace& trace) : writer_(trace)
+template <class Writer>
```

Comment:
No template implementation in cpp files

Reply:
https://isocpp.org/wiki/faq/templates#separate-template-fn-defn-from-decl The official CPP FAQ states that template instantiations belong into the cpp file. If I put it in the hpp, I am getting linker errors.

Reply:
This proves that C++ simply isn't teachable. And I should have been more precise in my remark. The FAQ predates C++11 and talks about `extern template`. No mainstream compiler ever supported that. Also, one item above, you can find this gem: https://isocpp.org/wiki/faq/templates#templates-defn-vs-decl

The general advice is, don't put template *definitions* in cpp files. Is there a correct way to do it? Yes. Should you do it? No. Just put template definitions in headers.

If you ever have to put a template definition in a cpp file, then the template instantiation also belongs there as well. That is what the FAQ talks about.